gnovm/pkg/gnolang/machine.go

```diff
@@ -224,35 +224,65 @@ func (m *Machine) SetActivePackage(pv *PackageValue) {
 //----------------------------------------
 // top level Run* methods.
 
-// Upon restart, preprocess all MemPackage and save blocknodes.
-// This is a temporary measure until we optimize/make-lazy.
+// PreprocessAllFilesAndSaveBlockNodes prepares the store for lazy loading
+// by registering packages in the store but not fully preprocessing them.
 //
 // NOTE: package paths not beginning with gno.land will be allowed to override,
 // to support cases of stdlibs processed through [RunMemPackagesWithOverrides].
 func (m *Machine) PreprocessAllFilesAndSaveBlockNodes() {
-	ch := m.Store.IterMemPackage()
-	for memPkg := range ch {
+	// Set up a PackageGetter that will do the actual preprocessing when needed
+	pkgGetter := func(pkgPath string, store Store) (*PackageNode, *PackageValue) {
+		// Get the original MemPackage
+		memPkg := store.GetMemPackage(pkgPath)
```

Comment:
Just a quick note: the problem with this is that nodes which are restarted have a different set of packages in the cache compared to ones which have been running with a cache; consequently, loading a DB from the store needs to happen with a clean gas meter.

I would also suggest that the best solution is one that keeps packages in the store using a cache, so that the amount of packages kept in the store is actually bounded to a specific amount

Reply:
I am thinking about alternative implementations because the one in the PR seems to have issues with persistence. I'll keep this in mind whether i continue with this or go in a different direction.